### PR TITLE
[WIP] Add support for building docker-engine RPMs for the Amazon Linux AMI

### PR DESCRIPTION
**Pull request by iliana**
This modifies `hack/make/build-rpm` to build RPMs for the Amazon Linux AMI. I'd greatly appreciate it if this could be considered for 1.9.

Because there is not a container image of the Amazon Linux AMI currently, the RPMs are built in a CentOS 7 container, which provides the most-similar version of RPM.

A side-effect of using the CentOS 7 container is the presence of systemd-libs-devel (or an equivalent), which causes Docker to link against libsystemd-journal. There's a hack added to `hack/make.sh` to disable the journald build tag — if you have a better suggestion than the change I've made, let me know.

I've done basic testing of a 1.9.0-dev RPM on the 2015.09 AMI, and verified that the CentOS 7 RPM still links to libsystemd-journal. I'd be happy to do thorough testing of officially built pre-release RPMs.

### CODE REVIEW

**Comment by calavera** [comment]
@jfrazelle can you take a look at this?

**Comment by calavera** [comment]
LGTM

**Comment by jfrazelle** [comment]
I assigned myself I'll take a look later

File: contrib/builder/rpm/generate.sh [hunk 1]
```diff
@@ -96,5 +109,10 @@ for version in "${versions[@]}"; do
 
 	echo 'ENV AUTO_GOPATH 1' >> "$version/Dockerfile"
 
-	echo 'ENV DOCKER_BUILDTAGS selinux' >> "$version/Dockerfile"
+	DOCKER_BUILDTAGS="selinux"
+	if [[ "$distro" == "amazon" ]]; then
+		DOCKER_BUILDTAGS="$DOCKER_BUILDTAGS !journald"
```
Comment (jfrazelle) [inline contrib/builder/rpm/generate.sh]: why?

Comment (ianweller) [inline contrib/builder/rpm/generate.sh]: The Amazon Linux AMI doesn't use systemd yet and doesn't have journald libraries, but the CentOS 7 container has the journald development libraries installed if device-mapper-devel is installed.

Comment (Djelibeybi) [inline contrib/builder/rpm/generate.sh]: Have you considering using the Oracle Linux 6 (`FROM oraclelinux:6`) image instead? That matches the Amazon Linux AMI far more closely than CentOS 7.

Comment (jfrazelle) [inline contrib/builder/rpm/generate.sh]: good call then we can avoid adding a new image :)

Comment (ianweller) [inline contrib/builder/rpm/generate.sh]: The main purpose of matching was to match based on what version of RPM and `rpmbuild` we have, which is closest to CentOS 7.

I'll look into changing the build image, although I'd strongly prefer the RPMs get built with a newer version of `rpmbuild`.

Comment (Djelibeybi) [inline contrib/builder/rpm/generate.sh]: In which case, it's probably easier just to build an `amazonlinux` Docker image, even if it's only used for building Docker for use on Amazon Linux AMIs.

**Comment by LK4D4** [comment]
ping @jfrazelle

File: contrib/builder/rpm/generate.sh (hunk 1 above)
Comment (jfrazelle) [inline contrib/builder/rpm/generate.sh]: +1 to amazon linux docker image

**Comment by LK4D4** [comment]
@ianweller This needs rebase :/
Also maybe @tianon can review.

**Comment by jfrazelle** [comment]
can we make an official amazon linux docker image for this, to build from

**Comment by thaJeztah** [comment]
ping @ianweller are you still interested in working on this? Not sure what the status is here. If we want this, perhaps a maintainer should carry this?

**Comment by jfrazelle** [comment]
I think we are waiting to see if they make an Amazon image

**Comment by thaJeztah** [comment]
I changed the title to [WIP], but perhaps we should close until there's an image

**Comment by ianweller** [comment]
Sorry for the delay in responding —

We don't have an Amazon Linux-based Docker image available at this time. We're hoping to make one available, but we don't have an estimated date right now.

So that leaves us with two options (I think):
* Continue working on a path where we can build RPMs for the Amazon Linux AMI, without having a container image, or
* Close this pull request and wait for that to happen.

**Comment by thaJeztah** [comment]
ping @jfrazelle what's your preference?

**Comment by jfrazelle** [comment]
can we wait for the image? is this a blocker without it, have you tried using the oracle linux 6 rpm

**Comment by ianweller** [comment]
Building RPMs for the Amazon Linux AMI isn't a blocker for anything, it's just a nice-to-have.

If the concern here is to not introduce the `DOCKER_BUILDTAGS="!journald"` bit, then I can probably use either the `centos:6` or `oraclelinux:6` containers, pending some testing on my end. (Is the preference for using `oraclelinux:6` because it can be reused, since packages aren't built for CentOS 6?)

(Was this closed on accident?)

**Comment by tianon** [comment]
If there's actually a non-zero chance of getting an Amazon Linux image,
then I'd personally prefer 100% that we wait (and think that such a thing
would be fantastically amazing -- so many +1 from me for that being a
thing).

Generally speaking, if there isn't an image for a particular cloud distro,
then I don't think it makes sense for the Docker project to officially
support it with packages that we ultimately won't be able to smoke-test
easily without spinning up real infra (in this case AWS is fairly cheap,
but the next thing that adding this might encourage may not be, and I don't
personally have ready access to AWS infra and am not real keen on paying
for some just to test this kind of thing and I'm pretty sure I'm not alone).

So, TL;DR, I'm really, really thrilled at the idea of an official Amazon
Linux base image! :+1:

**Comment by jfrazelle** [comment]
+1 @tianon :)

**Comment by Djelibeybi** [comment]
Are the Amazon Linux binary RPMs available publicly anywhere? If so, creating the image should be pretty easy. There are plenty of scripts that can do it already.

**Comment by redbaron** [comment]
FYI, Amazon released docker image with Amazon Linux: https://aws.amazon.com/blogs/aws/new-amazon-linux-container-image-for-cloud-and-on-premises-workloads/

**Comment by tianon** [comment]
And very soon that same image will be available directly as `amazonlinux` at the top-level: https://github.com/docker-library/official-images/pull/2314 :+1:

**Comment by lox** [comment]
Just trying to figure out the status of Docker 1.12 on Amazon Linux, what's the current blocker?
